actions/dump-charm-artifacts/action.yml

```diff
@@ -0,0 +1,52 @@
+name: 'Dump Charm Log Artifacts'
+description: 'Dumps to Artifacts useful Juju, Kubernetes, and Charmcraft log information.  Requires kubectl to be set up with a context, and charmcraft to be installed.'
+runs:
+  using: 'composite'
+  steps:
+    - name: Check out repo
+      uses: actions/checkout@v3
+
+    # Setup
+
+    - name: Create artifact collection directory
+      id: mkdir
+      shell: bash
+      if: always()
+      run: |
+        LOG_DIR="tmp_logs"
+        echo "log-dir=$(echo $LOG_DIR)" >> $GITHUB_OUTPUT
+        mkdir $LOG_DIR
+
+    - name: 'Install juju-crashdump'
+      shell: bash
+      if: always()
+      run: sudo snap install juju-crashdump --classic
+
+    - name: 'Install ketall'
+      # https://github.com/corneliusweig/ketall
+      shell: bash
+      if: always()
+      run: |
+        wget https://github.com/corneliusweig/ketall/releases/download/v1.3.8/get-all-amd64-linux.tar.gz
+        tar -xf get-all-amd64-linux.tar.gz
+        mkdir -p /home/runner/.local/bin/
+        mv get-all-amd64-linux /home/runner/.local/bin/ketall
+
+    # Dump logs
+
+    - name: Collect all logs
+      shell: bash
+      if: always()
+      run: |
+        export OUTPUT_DIR=${{ steps.mkdir.outputs.log-dir }}
+        bash ${{ github.action_path }}/logdump.bash
+
+    # Save logs
+
+    - name: Upload debug artifacts
+      uses: actions/upload-artifact@v3
+      if: always()
```

Comment:
If the previous step fails for whatever reason, we would have an empty artifact, are we sure we want to use `if: always()` here?

Reply:
If the previous step partly succeeds (say the last `ketall` command fails), we do have most of our logs even though it failed.  to me I'd rather still get some logs

but I wont fight on this if the preference is dropping `always()`.  I don't see the cost to having it but agree its an edge case at best

Reply:
counter idea

what if we update bash script that collects return codes but keeps running till the end.
at the end we raise the worst status code (or actually just increment it `return_code + return_code + return_code` will be 0 if all are good, or 3 if all 3 failed.

that will allow us to collect at maximum but not lose error message at the end (+ add debug prints say that this returned non zero)

Reply:
I like the idea, but I think it exceeds my bash knowledge :D

Reply:
I think I've achieved what @beliaev-maksim is after.  The script now traps errors and if any error happens, it will still try to do everything but will end with an error code.  

Also removed all `always()` except the last one for the upload.  That way if the logdump script exits with an error, it'll show :X: but still get logs uploaded.  

What does everyone think?

Reply:
I still see always()

or?

Reply:
sorry, messed something up on the push.  Should be good now

Reply:
yes, now is good